clash-cores/test/Test/Cores/LineCoding8b10b.hs

```diff
@@ -0,0 +1,115 @@
+
+-- | Recursive function to generate a list of 'Symbol8b10b's with the correct
+--   running disparity
+genSymbol8b10bs1 :: Int -> Bool -> H.Gen [(Bool, Symbol8b10b)]
+genSymbol8b10bs1 0 _ = pure []
+genSymbol8b10bs1 n rd = do
+  (rdNew, dw) <- genSymbol8b10b rd
+  ((rdNew, dw) :) <$> genSymbol8b10bs1 (pred n) rdNew
+
+-- | Generate a 'Symbol8b10b' by creating a 'BitVector' of length 10 and
+--   decoding it with the 'decode8b10b' function
+genSymbol8b10b :: Bool -> H.Gen (Bool, Symbol8b10b)
+genSymbol8b10b rd = Gen.filter f $ decode8b10b rd <$> genDefinedBitVector
+ where
+  f (_, dw) = isDw dw
+
+-- Check if the output of 'decode8b10b' is a valid value for a given value from
+-- 'encode8b10b'.
+prop_decode8b10bCheckNothing :: H.Property
+prop_decode8b10bCheckNothing = H.withTests 1000 $ H.property $ do
+  inp <- H.forAll genDefinedBitVector
+  let out = isValidSymbol dw1 && isValidSymbol dw2
+       where
+        (_, dw1) = decode8b10b False $ snd $ encode8b10b False (Dw inp)
+        (_, dw2) = decode8b10b True $ snd $ encode8b10b True (Dw inp)
+
+  H.assert out
+
+-- | Encode the input signal and check whether it is a valid value. It should be
+--   valid for every possible input.
+prop_encode8b10bCheckNothing :: H.Property
+prop_encode8b10bCheckNothing = H.withTests 1000 $ H.property $ do
+  inp <- H.forAll genDefinedBitVector
+  let out = 0 /= snd (encode8b10b False (Dw inp))
+
+  H.assert out
+
+-- | Encode and then decode the input signal, but if the result of the encode or
+--   decode functions is invalid, propagate the input itself to the output. The
+--   properties 'prop_decode8b10bCheckNothing' and
+--   'prop_encode8b10bCheckNothing' are used to assert that there are no invalid
+--   values in the outputs of these functions.
+prop_encodeDecode8b10b :: H.Property
+prop_encodeDecode8b10b = H.withTests 1000 $ H.property $ do
+  inp <- H.forAll genDefinedBitVector
+  let out = if isValidSymbol dw then fromDw dw else inp
+       where
+        dw = snd $ decode8b10b False $ snd $ encode8b10b False (Dw inp)
+
+      expected = inp
+
+  out H.=== expected
+
+-- | Encode and then decode the input signal, but if the result of the encode or
+--   decode functions is invalid, propagate the input itself to the output as
+--   the set of valid code groups is much larger than the set of valid data
+--   words. The generated list is filtered to make sure that there are no
+--   accidental commas in the input list, as these are only accepted for a small
+--   subset of control signals.
+prop_decodeEncode8b10b :: H.Property
+prop_decodeEncode8b10b = H.withTests 1000 $ H.property $ do
+  inp <-
+    H.forAll
+      (Gen.filter checkBitSequence genDefinedBitVector)
+  let out = if o == inp then o else g True inp
+       where
+        o = g False inp
+
+        g rd i = if isValidSymbol dw then snd $ encode8b10b rd dw else i
+         where
+          dw = snd $ decode8b10b rd i
+
+      expected = inp
+
+  out H.=== expected
```

Comment:
I thought this test was flawed. If `encode8b18b False dw` did _not_ result in a valid data word, we don't try it with `encode8b10b True dw`, we just immediately return success. So I rewrote it to:
```suggestion
  let out = case (o1valid, o2valid) of
              (True, _) -> o1
              (_, True) -> o2
              _ -> inp
       where
        (o1valid, o1) = g False inp
        (o2valid, o2) = g True inp

        g rd i = (isValidSymbol dw, snd $ encode8b10b rd dw)
         where
          dw = snd $ decode8b10b rd i

      expected = inp

  out H.=== expected
```

but now I got counterexamples. They truly are counterexamples. For instance, the code 0b11_0000_1001 represents D16.3 if the running disparity is `True`, but the code 0b00_1100_1001 also represents D16.3 under the same conditions. So we get:

```
    decodeEncode8b10b:       FAIL (0.02s)
        ✗ <interactive> failed at test/Test/Cores/LineCoding8b10b.hs:110:7
          after 11 tests and 3 shrinks.
          shrink path: 11:mBa

              ┏━━ test/Test/Cores/LineCoding8b10b.hs ━━━
           91 ┃ prop_decodeEncode8b10b :: H.Property
           92 ┃ prop_decodeEncode8b10b = H.withTests 1000 $ H.property $ do
           93 ┃   inp <-
           94 ┃     H.forAll
              ┃     │ 0 b11_0000_1001
           95 ┃       (Gen.filter checkBitSequence genDefinedBitVector)
           96 ┃   let out = case (o1valid, o2valid) of
           97 ┃               (True, _) -> o1
           98 ┃               (_, True) -> o2
           99 ┃               _ -> inp
          100 ┃        where
          101 ┃         (o1valid, o1) = g False inp
          102 ┃         (o2valid, o2) = g True inp
          103 ┃
          104 ┃         g rd i = (isValidSymbol dw, snd $ encode8b10b rd dw)
          105 ┃          where
          106 ┃           dw = snd $ decode8b10b rd i
          107 ┃
          108 ┃       expected = inp
          109 ┃
          110 ┃   out H.=== expected
              ┃   ^^^^^^^^^^^^^^^^^^
              ┃   │ ━━━ Failed (- lhs) (+ rhs) ━━━
              ┃   │ - 0 b00_1100_1001
              ┃   │ + 0 b11_0000_1001
```

because it picked the other alternative for D16.3.

So let's discuss how we should improve the test.

Reply:
Interesting, I will look at this tomorrow!

Reply:
Alright, I decided to add an extra decoding step for both the input and the output to be able to check whether they are the same without having to deal with the running disparity. Does this look OK?

Reply:
I think the test as I suggested is correct but that the table returns `Dw`s when in fact they are errors. By hand I now think I found these 5:
```diff
--- a/clash-cores/src/Clash/Cores/LineCoding8b10b/Decoder.hs
+++ b/clash-cores/src/Clash/Cores/LineCoding8b10b/Decoder.hs
@@ -1495,7 +1495,7 @@ decoderLut =
   , (0, 0, 0, 1, 0b11111100) -- D28.7
   , (1, 1, 0, 1, 0b11100000) -- D00.7
   , (0, 0, 0, 1, 0b11101111) -- D15.7
-  , (0, 0, 0, 1, 0b11100000) -- D00.7
+  , (0, 1, 0, 1, 0b11100000) -- D00.7
   , (1, 0, 0, 0, 0b11100111) -- D07.7
   , (1, 1, 0, 1, 0b11100000) -- D00.7
   , (0, 0, 0, 1, 0b11110000) -- D16.7
@@ -1814,11 +1814,11 @@ decoderLut =
   , (1, 1, 0, 0, 0b01100000) -- D00.3
   , (0, 0, 0, 0, 0b01111100) -- D28.3
   , (1, 1, 0, 0, 0b01100000) -- D00.3
-  , (0, 0, 0, 0, 0b01101111) -- D15.3
-  , (0, 0, 0, 0, 0b01100000) -- D00.3
+  , (0, 1, 0, 0, 0b01101111) -- D15.3
+  , (0, 1, 0, 0, 0b01100000) -- D00.3
   , (0, 0, 0, 1, 0b01100111) -- D07.3
   , (1, 1, 0, 0, 0b01100000) -- D00.3
-  , (0, 0, 0, 0, 0b01110000) -- D16.3
+  , (0, 1, 0, 0, 0b01110000) -- D16.3
   , (0, 0, 0, 0, 0b01111111) -- D31.3
   , (0, 0, 0, 1, 0b01101011) -- D11.3
   , (0, 0, 0, 0, 0b01111000) -- D24.3
@@ -1942,7 +1942,7 @@ decoderLut =
   , (1, 1, 0, 1, 0b11100000) -- D00.7
   , (0, 0, 1, 1, 0b11111100) -- K28.7
   , (1, 1, 0, 1, 0b11100000) -- D00.7
-  , (0, 0, 0, 1, 0b11101111) -- D15.7
+  , (0, 1, 0, 1, 0b11101111) -- D15.7
   , (0, 0, 0, 1, 0b11100000) -- D00.7
   , (1, 0, 0, 0, 0b11100111) -- D07.7
   , (1, 1, 0, 1, 0b11100000) -- D00.7
```

but this is not fun by hand and easily automated.

However, you were trying to be quickly done with this part of the work to concentrate on the interesting stuff instead, and now we're spending time on this part anyway. The errors I found were all non-critical.¹ So I think I'll suggest merging as-is, and making an issue that says a) this can be so much less resource-intensive and b) it appears to have mistakes, so it should just be replaced by a better implementation (not to be part of your internship, unless of course you find some extra time, but that is not something that happens in an internship I think).

¹ It will decode _some_ garbage as valid data, which is "be liberal in what you accept" taken too far, but in practice will not quickly lead to issues.